fix(aio): improve search index to capture better results

petebacondarwin opened the pull request:
Fixes #17545 #17889

petebacondarwin commented:
Before: 
![screen shot 2017-06-16 at 20 26 56](https://user-images.githubusercontent.com/15655/27241822-41a6cf12-52d2-11e7-9510-55817dfede77.png)

petebacondarwin commented:
After:

![screen shot 2017-06-16 at 20 27 10](https://user-images.githubusercontent.com/15655/27241828-4752b598-52d2-11e7-8c7a-582957a3e706.png)

mary-poppins commented:
The angular.io preview for 71da637f66d47351bacd9e71918f8bf6e49c5b10 is available [here][1].

[1]: https://pr17574-71da637f66d47351bacd9e71918f8bf6e49c5b10.ngbuilds.io/

gkalpak commented on aio/src/app/search/search-worker.js:
```diff
@@ -86,8 +86,11 @@ function queryIndex(query) {
+      // Allow fuzzy matching to improve results
       if (terms.length === 1 || term.trim().length > 2) {
+        qb.term(term, { usePipeline: true });
         qb.term(term, { wildcard: lunr.Query.wildcard.TRAILING });
```
Would  it help fnd more relevant results if we used a leading `*` too?

gkalpak reviewed (commented):
I feel we are getting too many results now :grin: (but maybe I am wrong). For example the term "Date" gives `Meta` as the first result, because it contains an `updateXyz` method.

I don't know what lunr's capabilities are and if we can get more accurate ranking 😕 

##
Also, I am not sure this fixes the underlying cause of #17545. Why weren't we previously matching the `Validator` interface to the "validator" term (they match 1-1)?

petebacondarwin commented:
How about we run the previous search and then this more relaxed search if
the results list is small?

On Sat, 17 Jun 2017 at 07:37 George Kalpakas <notifications@github.com>
wrote:

> *@gkalpak* commented on this pull request.
>
> I feel we are getting too many results now 😁 (but maybe I am wrong). For
> example the term "Date" gives Meta as the first result, because it
> contains an updateXyz method.
>
> I don't know what lunr's capabilities are and if we can get more accurate
> ranking 😕
>
> Also, I am not sure this fixes the underlying cause of #17545
> <https://github.com/angular/angular/issues/17545>. Why weren't we
> previously matching the Validator interface to the "validator" term (they
> match 1-1)?
> ------------------------------
>
> In aio/src/app/search/search-worker.js
> <https://github.com/angular/angular/pull/17574#discussion_r122562386>:
>
> >        if (terms.length === 1 || term.trim().length > 2) {
> +        qb.term(term, { usePipeline: true });
>          qb.term(term, { wildcard: lunr.Query.wildcard.TRAILING });
>
> Would it help fnd more relevant results if we used a leading * too?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/17574#pullrequestreview-44694457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA9J7YQwUcNRbyM-O2lKL7vZjRkGM80ks5sE3RGgaJpZM4N8186>
> .
>

gkalpak commented:
Sounds promising :smiley:

petebacondarwin committed:
build(aio): revert to general purpose search algorithm

Now that we have upgraded to the latest lunr search engine, the results
from the standard `search` method are more appropriate.
So we do not need to create our own special queries to get good results.

petebacondarwin committed:
build(aio): capture all the headings from a doc in the vFile.headings property

petebacondarwin committed:
build(aio): add terms from heading to the search index

mary-poppins commented:
You can preview 7a6081c at https://pr17574-7a6081c.ngbuilds.io/.

petebacondarwin commented:
The last improvement needed to get things like the `ngModules` guide to the top of the results is a bug fix in lunr: https://github.com/olivernn/lunr.js/pull/281

But we should merge this PR for now and then update the Lunr dependency when @olivernn lands that fix to Lunr.

gkalpak reviewed (approved):
This is obviously much better than before, but I still find the behavior a little odd. For example as you type `animations`, results are shown and hidden: 

- `anim` -> Many results
- `anima` -> No results
- `animat` -> One result
- `animati` -> No results
- `animatio` -> No results
- `animation` -> Many results

I assume this is not something that olivernn/lunr.js#281 would fix, right?

##
This PR is a very good step forward anyway though. So 👍

petebacondarwin commented:
> For example as you type animations, results are shown and hidden:
> 
> anim -> Many results
> anima -> No results
> animat -> One result
> animati -> No results
> animatio -> No results
> animation -> Many results

This is not fixed by the changes to lunr.js. It is related to how the stemmer works. The actual queries will look a bit like this:

anim -> anim(ate)
anima -> anima
animat -> animat(e)
animati -> animati
animatio -> animatio
animation -> animat(ion)

I was thinking that a bit of debounce (even 200ms) would improve the UX as the query will only happen after the user finishes typing and so will hide these unfortunate results.

olivernn commented:
The problem is due to the stemming of the terms, you can _probably_ get better results by constructing the query slightly differently, searching with a wildcard as well as exact matches.

Something like the following would probably work:

```javascript
idx.query(function (q) {
  q.term("anima", { boost: 100 })  // exact match
  q.term("anima", { boost: 10, wildcard: lunr.Query.wildcard.TRAILING, usePipeline: false }) // prefix search, no stemming
  q.term("anima", { editDistance: 1, usePipeline: false }) // fuzzy search, no stemming
})
```

You can try this out on the demo by searching for "anima anima* anima~1". The search with edit distance is optional but will increase recall.

petebacondarwin commented:
@olivernn I did originally try building my own queries, using a similar mix of wildcards and stemming, etc, but in the end I found that your query parser does a better job than I could.

petebacondarwin commented:
Just waiting for Travis to approve.

mary-poppins commented:
You can preview bd9f5c6 at https://pr17574-bd9f5c6.ngbuilds.io/.

mary-poppins commented:
You can preview 27b4cc7 at https://pr17574-27b4cc7.ngbuilds.io/.

petebacondarwin commented:
OK, so this PR needs to sit on top of the #17857

mary-poppins commented:
You can preview f61edbe at https://pr17574-f61edbe.ngbuilds.io/.

Foxandxss commented:
Guide changes are good for me.

petebacondarwin commented:
I need to wait for #17857 to land and then rebase before we can land this.

mary-poppins commented:
You can preview 6d8fe54 at https://pr17574-6d8fe54.ngbuilds.io/.

petebacondarwin commented:
Cherry picked to 4.2.x as 

dbfd7179a8fb3c44ffc64688b12ab72bbc41a797 build(aio): add terms from heading to the search index
207e2febb302b6ed372a6dfe4d2c6245ae9c9d5b build(aio): capture all the headings from a doc in the vFile.headings property
39c9860b9d4fdbca4218de7989b59171956e92da build(aio): revert to general purpose search algorithm
